src/Compilers/CSharp/Portable/Compilation/MemberSemanticModel.cs

```diff
@@ -1384,29 +1384,70 @@ private static Binder GetQueryEnclosingBinder(int position, BoundQueryClause que
+                            {
+                                n = arg;
+                                break;
+                            }
+                        }
+
+                        if (n != call)
+                        {
+                            continue;
+                        }
+
+                        BoundExpression receiver = call.ReceiverOpt;
+
+                        // In some error scenarios, we end-up with a method group as the receiver,
+                        // let's get to real receiver.
+                        while (receiver?.Kind == BoundKind.MethodGroup)
+                        {
+                            receiver = ((BoundMethodGroup)receiver).ReceiverOpt;
                         }
+
+                        if (receiver?.Syntax.FullSpan.Contains(position) == true ||
+                            (receiver as BoundQueryClause)?.Value.Syntax.FullSpan.Contains(position) == true)
+                        {
+                            n = receiver;
+                            continue;
+                        }
+
+                        // TODO: should we look for the "nearest" argument as a fallback?
+                        n = call.Arguments.LastOrDefault();
                         continue;
                     case BoundKind.Conversion:
                         n = ((BoundConversion)n).Operand;
                         continue;
                     case BoundKind.UnboundLambda:
-                        // NOTE: Calling GetLambdaEnclosingBinder would just return this binder.
-                        return ((UnboundLambda)n).BindForErrorRecovery().Binder;
+                        var unbound = (UnboundLambda)n;
+                        return GetEnclosingBinder(AdjustStartingNodeAccordingToNewRoot(startingNode, unbound.Syntax),
+                                                  position, unbound.BindForErrorRecovery().Binder, unbound.Syntax);
                     case BoundKind.Lambda:
-                        // NOTE: Calling GetLambdaEnclosingBinder would just return this binder.
-                        return ((BoundLambda)n).Binder;
+                        var lambda = (BoundLambda)n;
+                        return GetEnclosingBinder(AdjustStartingNodeAccordingToNewRoot(startingNode, lambda.Body.Syntax), 
+                                                  position, lambda.Binder, lambda.Body.Syntax);
                     default:
-                        return queryClause.Binder;
+                        goto done;
                 }
             }
 
-            return queryClause.Binder;
+done:
+            return GetEnclosingBinder(AdjustStartingNodeAccordingToNewRoot(startingNode, queryClause.Syntax),
+                                      position, queryClause.Binder, queryClause.Syntax);
+        }
+
+        private static CSharpSyntaxNode AdjustStartingNodeAccordingToNewRoot(CSharpSyntaxNode startingNode, CSharpSyntaxNode root)
```

Comment:
Yes, I am going to open a separate issue to overhaul the way we look for binders in queries. It is too fragile and not as good with error recovery as lambdas.